In [2]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [54]:
csv_file = "../data/spotify_dataset.csv"
spotify_df = pd.read_csv(csv_file, sep=",")
print(f"Dataset shape: {spotify_df.shape[0]} Rows and {spotify_df.shape[1]} Columns")
spotify_df.head()

Dataset shape: 114000 Rows and 21 Columns


Unnamed: 0                track_id                 artists  \
0           0  5SuOikwiRyPMVoIQDJUgSV             Gen Hoshino   
1           1  4qPNDBW1i3p13qLCt0Ki3A            Ben Woodward   
2           2  1iJBSr7s7jYXzM8EGcbK5b  Ingrid Michaelson;ZAYN   
3           3  6lfxq3CG4xtTiEg7opyCyx            Kina Grannis   
4           4  5vjLSffimiIP26QG5WcN2K        Chord Overstreet   

                                          album_name  \
0                                             Comedy   
1                                   Ghost (Acoustic)   
2                                     To Begin Again   
3  Crazy Rich Asians (Original Motion Picture Sou...   
4                                            Hold On   

                   track_name  popularity  duration_ms  explicit  \
0                      Comedy          73       230666     False   
1            Ghost - Acoustic          55       149610     False   
2              To Begin Again          57       210826     False   
3  Can't Help Falling In Love          71       201933     False   
4                     Hold On          82       198853     False   

   danceability  energy  ...  loudness  mode  speechiness  acousticness  \
0         0.676  0.4610  ...    -6.746     0       0.1430        0.0322   
1         0.420  0.1660  ...   -17.235     1       0.0763        0.9240   
2         0.438  0.3590  ...    -9.734     1       0.0557        0.2100   
3         0.266  0.0596  ...   -18.515     1       0.0363        0.9050   
4         0.618  0.4430  ...    -9.681     1       0.0526        0.4690   

   instrumentalness  liveness  valence    tempo  time_signature  track_genre  
0          0.000001    0.3580    0.715   87.917               4     acoustic  
1          0.000006    0.1010    0.267   77.489               4     acoustic  
2          0.000000    0.1170    0.120   76.332               4     acoustic  
3          0.000071    0.1320    0.143  181.740               3     acoustic  
4          0.000000    0.0829    0.167  119.949               4     acoustic  

[5 rows x 21 columns]

In [43]:
spotify_df.columns

Index(['Unnamed: 0', 'track_id', 'artists', 'album_name', 'track_name',
       'popularity', 'duration_ms', 'explicit', 'danceability', 'energy',
       'key', 'loudness', 'mode', 'speechiness', 'acousticness',
       'instrumentalness', 'liveness', 'valence', 'tempo', 'time_signature',
       'track_genre'],
      dtype='object')

#### Columns:

- **track_id:** Spotify ID for the track.

- **artists:** Names of the track's performing artists (semicolon-separated if multiple).
- **album_name:** Name of the album containing the track.
- **track_name:** Name of the track.
- **popularity:** A measure of track popularity, ranging from 0 to 100.
- **duration_ms:** Track length in milliseconds.
- **explicit:** Indicates if the track has explicit lyrics (true = yes, false = no or unknown).
- **danceability:** Describes how suitable the track is for dancing, with values between 0.0 (least danceable) and 1.0 (most danceable).
- **energy:** A measure of intensity and activity in the track, ranging from 0.0 to 1.0.
- **key:**  The musical key of the track, mapped to standard Pitch Class notation (e.g., 0 = C, 1 = C♯/D♭).
- **loudness:** Overall loudness of the track in decibels (dB).
- **mode:** Indicates the modality (major or minor) of the track's melodic content (1 for major, 0 for minor).
- **speechiness:** Detects the presence of spoken words in the track, with values from 0.0 to 1.0.
- **acousticness:** A confidence measure of whether the track is acoustic (1.0 for high confidence).
- **instrumentalness:**  Predicts if the track contains no vocals, with values closer to 1.0 indicating no vocal content.
- **liveness:** Detects the presence of an audience in the recording, with values above 0.8 indicating a high likelihood of a live performance.
- **valence:** A measure from 0.0 to 1.0 describing the musical positiveness conveyed by the track.
- **tempo:** Estimated tempo of the track in beats per minute (BPM).
- **time_signature:** Estimated time signature, ranging from 3 to 7 (e.g., 3/4 to 7/4).
- **track_genre:** The genre to which the track belongs.

In [44]:
spotify_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114000 entries, 0 to 113999
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Unnamed: 0        114000 non-null  int64  
 1   track_id          114000 non-null  object 
 2   artists           113999 non-null  object 
 3   album_name        113999 non-null  object 
 4   track_name        113999 non-null  object 
 5   popularity        114000 non-null  int64  
 6   duration_ms       114000 non-null  int64  
 7   explicit          114000 non-null  bool   
 8   danceability      114000 non-null  float64
 9   energy            114000 non-null  float64
 10  key               114000 non-null  int64  
 11  loudness          114000 non-null  float64
 12  mode              114000 non-null  int64  
 13  speechiness       114000 non-null  float64
 14  acousticness      114000 non-null  float64
 15  instrumentalness  114000 non-null  float64
 16  liveness          11

In [45]:
spotify_df.isnull().sum()

Unnamed: 0          0
track_id            0
artists             1
album_name          1
track_name          1
popularity          0
duration_ms         0
explicit            0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
time_signature      0
track_genre         0
dtype: int64

We can see that the artist, album name and track name have one null value each, are these null values in the same row? If that's the case, we might have to drop that row eventually.

In [46]:
rows_with_nulls = spotify_df[spotify_df.isnull().any(axis=1)]
print(rows_with_nulls)

       Unnamed: 0                track_id artists album_name track_name  \
65900       65900  1kR4gIb7nGxHPI3D2ifs59     NaN        NaN        NaN   

       popularity  duration_ms  explicit  danceability  energy  ...  loudness  \
65900           0            0     False         0.501   0.583  ...     -9.46   

       mode  speechiness  acousticness  instrumentalness  liveness  valence  \
65900     0       0.0605          0.69           0.00396    0.0747    0.734   

         tempo  time_signature  track_genre  
65900  138.391               4        k-pop  

[1 rows x 21 columns]


We can see that, in fact, it's a single row that contains null values for the aforementioned columns.

In [47]:
spotify_df.nunique()

Unnamed: 0          114000
track_id             89741
artists              31437
album_name           46589
track_name           73608
popularity             101
duration_ms          50697
explicit                 2
danceability          1174
energy                2083
key                     12
loudness             19480
mode                     2
speechiness           1489
acousticness          5061
instrumentalness      5346
liveness              1722
valence               1790
tempo                45653
time_signature           5
track_genre            114
dtype: int64

In [48]:
spotify_df.describe()

Unnamed: 0     popularity   duration_ms   danceability  \
count  114000.000000  114000.000000  1.140000e+05  114000.000000   
mean    56999.500000      33.238535  2.280292e+05       0.566800   
std     32909.109681      22.305078  1.072977e+05       0.173542   
min         0.000000       0.000000  0.000000e+00       0.000000   
25%     28499.750000      17.000000  1.740660e+05       0.456000   
50%     56999.500000      35.000000  2.129060e+05       0.580000   
75%     85499.250000      50.000000  2.615060e+05       0.695000   
max    113999.000000     100.000000  5.237295e+06       0.985000   

              energy            key       loudness           mode  \
count  114000.000000  114000.000000  114000.000000  114000.000000   
mean        0.641383       5.309140      -8.258960       0.637553   
std         0.251529       3.559987       5.029337       0.480709   
min         0.000000       0.000000     -49.531000       0.000000   
25%         0.472000       2.000000     -10.013000       0.000000   
50%         0.685000       5.000000      -7.004000       1.000000   
75%         0.854000       8.000000      -5.003000       1.000000   
max         1.000000      11.000000       4.532000       1.000000   

         speechiness   acousticness  instrumentalness       liveness  \
count  114000.000000  114000.000000     114000.000000  114000.000000   
mean        0.084652       0.314910          0.156050       0.213553   
std         0.105732       0.332523          0.309555       0.190378   
min         0.000000       0.000000          0.000000       0.000000   
25%         0.035900       0.016900          0.000000       0.098000   
50%         0.048900       0.169000          0.000042       0.132000   
75%         0.084500       0.598000          0.049000       0.273000   
max         0.965000       0.996000          1.000000       1.000000   

             valence          tempo  time_signature  
count  114000.000000  114000.000000   114000.000000  
mean        0.474068     122.147837        3.904035  
std         0.259261      29.978197        0.432621  
min         0.000000       0.000000        0.000000  
25%         0.260000      99.218750        4.000000  
50%         0.464000     122.017000        4.000000  
75%         0.683000     140.071000        4.000000  
max         0.995000     243.372000        5.000000

In [49]:
spotify_df.describe(include=["bool", "object"])

track_id      artists                  album_name  \
count                   114000       113999                      113999   
unique                   89741        31437                       46589   
top     6S3JlDAGk3uu3NtZbPnuhS  The Beatles  Alternative Christmas 2022   
freq                         9          279                         195   

             track_name explicit track_genre  
count            113999   114000      114000  
unique            73608        2         114  
top     Run Rudolph Run    False    acoustic  
freq                151   104253        1000

Why are there repeated songs in this dataset? Let's look more into that...

In [55]:
spotify_df.sort_values(by = 'track_name').head(10)

Unnamed: 0                track_id  \
36750        36750  0fROT4kK5oTm8xO8PX6EJF   
92751        92751  1hH0t381PIXmUVWyG1Vj3p   
66970        66970  1B45DvGMoFWdbAEUH2qliG   
66996        66996  73lXf5if6MWVWnsgXhK8bd   
101161      101161  0jmz4aHEIBCRgrcV2xEkwB   
82584        82584  5Zx0Rrkn5RFBMD2PRxX3mI   
5980          5980  3ozivYJGJGq6TSzdy8m64X   
93440        93440  3KKk48f33mlB56F5L5nbJk   
93397        93397  3KKk48f33mlB56F5L5nbJk   
93895        93895  5OiONTndVC5YOMXg6VC5xs   

                                                  artists  \
36750                                               Rilès   
92751                                        Brian Hyland   
66970                                   Little Apple Band   
66996                                   Little Apple Band   
101161  Traditional;Sistine Chapel Choir;Massimo Palom...   
82584                                      Dillinger Four   
5980                                    Capcom Sound Team   
93440                                     Nikolay Kopylov   
93397                                     Nikolay Kopylov   
93895                                     Nikolay Kopylov   

                                 album_name  \
36750                        !I'll Be Back!   
92751                     The Bashful Blond   
66970   The Favorite Songs Of Sesame Street   
66996             Sesame Street and Friends   
101161                  Classical Christmas   
82584                       C I V I L W A R   
5980            デビル メイ クライ 3 オリジナル・サウンドトラック   
93440                   Popular Opera Arias   
93397                   Popular Opera Arias   
93895                   Popular Opera Arias   

                                    track_name  popularity  duration_ms  \
36750                           !I'll Be Back!          52       178533   
92751                      "A" You're Adorable          39       151680   
66970                        "C" IS FOR COOKIE          32        84305   
66996                        "C" is for Cookie           8        86675   
101161             "Christe, Redemptor omnium"           0       289133   
82584   "Contemplate This on the Tree of Woe."          24       180706   
5980               "DEVILS NEVER CRY"(スタッフロール)          55       319906   
93440       "Don Carlos" Roderigo'S Death Aria           0       235547   
93397       "Don Carlos" Roderigo'S Death Aria           0       235547   
93895         "Eugene Onegin" Ariozo Of Onegin           0       111800   

        explicit  danceability  energy  ...  loudness  mode  speechiness  \
36750       True         0.823  0.6120  ...    -7.767     1       0.2480   
92751      False         0.615  0.3750  ...   -10.362     0       0.0319   
66970      False         0.553  0.8120  ...    -5.542     1       0.0558   
66996      False         0.664  0.6110  ...    -8.687     1       0.0886   
101161     False         0.111  0.0568  ...   -28.053     1       0.0551   
82584      False         0.565  0.9770  ...    -3.592     1       0.0546   
5980       False         0.264  0.9510  ...    -7.356     1       0.1460   
93440      False         0.167  0.3320  ...    -9.888     1       0.0359   
93397      False         0.167  0.3320  ...    -9.888     1       0.0359   
93895      False         0.443  0.5140  ...    -8.068     1       0.0924   

        acousticness  instrumentalness  liveness  valence    tempo  \
36750       0.168000           0.00000    0.1090   0.6880  142.959   
92751       0.482000           0.00000    0.1110   0.9220  110.720   
66970       0.132000           0.00001    0.0794   0.8710  118.368   
66996       0.120000           0.00000    0.0408   0.7580  118.443   
101161      0.990000           0.69700    0.1100   0.0395  169.401   
82584       0.028200           0.00132    0.4330   0.3870  106.478   
5980        0.000894           0.04420    0.1270   0.1590  149.990   
93440       0.992000           0.19100    0.1130   0.0527   62.970   
93397       0.9920

We can see two songs with the same title but one is in lower case and the other is in upper case. Despite the fact that these two songs might be the same, they come from different albums so I will treat them as different tracks.

In [60]:
spotify_df.duplicated(subset=['track_name','album_name','artists']).sum()

24620

There are 24620 duplicated songs in the dataset.

In [63]:
grouped_df = spotify_df.groupby(['track_name','album_name','artists']).size().reset_index(name='count') # grouping by artist and track name
grouped_df.sort_values(by = 'count', ascending= False) # sorting descending

track_name  \
40094               Last Christmas   
13391                Christmastime   
34575                    If We Try   
6458   Baby Blue - Remastered 2010   
68796                      Survive   
...                            ...   
32921   I Get The Sweetest Feeling   
32920                 I Get Looser   
32919                  I Get Loose   
32918      I Get Lonely In A Hurry   
89378                      행복하길 바래   

                                           album_name                artists  \
40094                      Alternative Christmas 2022        Jimmy Eat World   
13391                      Alternative Christmas 2022  The Smashing Pumpkins   
34575                             Best 70s Rock Tunes             Don McLean   
6458   Straight Up (Remastered 2010 / Deluxe Edition)              Badfinger   
68796                             Best 70s Rock Tunes          Jimmy Buffett   
...                                               ...                    ...   
32921                      I Get The Sweetest Feeling          Jackie Wilson   
32920                        Cafetorium Songs, Vol. 2      Koo Koo Kanga Roo   
32919                        Cafetorium Songs, Vol. 1      Koo Koo Kanga Roo   
32918                         I Get Lonely In A Hurry           George Jones   
89378    Sassy Girl Choon Hyang (Original Soundtrack)          Lim Hyung Joo   

       count  
40094     12  
13391     10  
34575      9  
6458       9  
68796      9  
...      ...  
32921      1  
32920      1  
32919      1  
32918      1  
89378      1  

[89379 rows x 4 columns]

There is a total of 89379 unique songs. The one that repeats the most appears 12 times in the dataset.

In [66]:
popular_songs_df = spotify_df.groupby(['track_name','album_name','artists'])['popularity'].idxmax() # grouping by track name, artist and album name and choosing the row with the highest popularity

most_popular_songs = spotify_df.loc[popular_songs_df] # rows with the song with the highest popularity

most_popular_songs

Unnamed: 0                track_id  \
36750        36750  0fROT4kK5oTm8xO8PX6EJF   
92751        92751  1hH0t381PIXmUVWyG1Vj3p   
66970        66970  1B45DvGMoFWdbAEUH2qliG   
66996        66996  73lXf5if6MWVWnsgXhK8bd   
101161      101161  0jmz4aHEIBCRgrcV2xEkwB   
...            ...                     ...   
65756        65756  2oVHb8wyg6oC2iNpGBNvx1   
65859        65859  4kIpBfvK44bxqX7zo8K1oP   
75903        75903  4mHc7LUlO3k6AXeFV2EiJK   
79554        79554  0tQesiSZJQOdHeAC7r59us   
76400        76400  5epH8xMDeJ4edJjIkKzEdM   

                                                  artists  \
36750                                               Rilès   
92751                                        Brian Hyland   
66970                                   Little Apple Band   
66996                                   Little Apple Band   
101161  Traditional;Sistine Chapel Choir;Massimo Palom...   
...                                                   ...   
65756                                               HEIZE   
65859                                                Gaho   
75903                                              Yiruma   
79554                                                GODA   
76400                                       Lim Hyung Joo   

                                               album_name  \
36750                                      !I'll Be Back!   
92751                                   The Bashful Blond   
66970                 The Favorite Songs Of Sesame Street   
66996                           Sesame Street and Friends   
101161                                Classical Christmas   
...                                                   ...   
65756   Hotel del Luna (Original Television Soundtrack...   
65859   ITAEWON CLASS (Original Television Soundtrack)...   
75903   Yiruma Official Album 'Piano Therapy' (The Ori...   
79554                 One Punch Man (Original Soundtrack)   
76400        Sassy Girl Choon Hyang (Original Soundtrack)   

                         track_name  popularity  duration_ms  explicit  \
36750                !I'll Be Back!          52       178533      True   
92751           "A" You're Adorable          39       151680     False   
66970             "C" IS FOR COOKIE          32        84305     False   
66996             "C" is for Cookie           8        86675     False   
101161  "Christe, Redemptor omnium"           0       289133     False   
...                             ...         ...          ...       ...   
65756                  내 맘을 볼 수 있나요          63       225785     False   
65859                            시작          65       202440     False   
75903               약속 (Piano Solo)          37       144533     False   
79554         원펀맨 Theme - Sad Theme          37       228000     False   
76400                       행복하길 바래          39       185666     False   

        danceability  energy  ...  loudness  mode  speechiness  acousticness  \
36750          0.823  0.6120  ...    -7.767     1       0.2480         0.168   
92751          0.615  0.3750  ...   -10.362     0       0.0319         0.482   
66970          0.553  0.8120  ...    -5.542     1       0.0558         0.132   
66996          0.664  0.6110  ...    -8.687     1       0.0886         0.120   
101161         0.111  0.0568  ...   -28.053     1       0.0551         0.990   
...              ...     ...  ...       ...   ...          ...           ...   
65756          0.397  0.1660  ...   -10.678     1       0.0349         0.884   
65859          0.591  0.8180  ...    -3.532     1       0.0730         0.172   
75903          0.504  0.1430  ...   -18.713     1       0.0701         0.980   
79554          0.542  0.3370  ...   -16.088     1       0.0539         0.985   
76400          0.361  0.5170  ...    -6.955     0       0.0310         0.711   

        instrumentalness  liveness  valence    tempo  time_signature  \
36750           0.000000    0.1090   0.6880  142.959               

In [ ]:
# TODO: corrplot